src/rgw/rgw_sync_log_trim.cc

```diff
@@ -335,6 +336,150 @@ int accumulate_peer_counters(bufferlist& bl, BucketChangeCounter& counter)
+using MetadataListCallback = std::function<bool(std::string&&, std::string&&)>;
+
+/// lists metadata keys, passing each to a callback until it returns false.
+/// on reaching the end, it will restart at the beginning and list up to the
+/// initial marker
+class AsyncMetadataList : public RGWAsyncRadosRequest {
+  CephContext *const cct;
+  RGWMetadataManager *const mgr;
+  const std::string section;
+  const std::string start_marker;
+  MetadataListCallback callback;
+  void *handle{nullptr};
+
+  int _send_request() override;
+ public:
+  AsyncMetadataList(CephContext *cct, RGWCoroutine *caller,
+                    RGWAioCompletionNotifier *cn, RGWMetadataManager *mgr,
+                    const std::string& section, const std::string& start_marker,
+                    const MetadataListCallback& callback)
+    : RGWAsyncRadosRequest(caller, cn), cct(cct), mgr(mgr),
+      section(section), start_marker(start_marker), callback(callback)
+  {}
+  ~AsyncMetadataList() override {
+    if (handle) {
+      mgr->list_keys_complete(handle);
+    }
+  }
+};
+
+int AsyncMetadataList::_send_request()
+{
+  // start a listing at the given marker
+  int r = mgr->list_keys_init(section, start_marker, &handle);
+  if (r < 0) {
+    ldout(cct, 10) << "failed to init metadata listing: "
+        << cpp_strerror(r) << dendl;
+    return r;
+  }
+  ldout(cct, 20) << "starting metadata listing at " << start_marker << dendl;
+
+  std::list<std::string> keys;
+  bool truncated{false};
+  std::string marker;
+
+  do {
+    // get the next key and marker
+    r = mgr->list_keys_next(handle, 1, keys, &truncated);
+    if (r < 0) {
+      ldout(cct, 10) << "failed to list metadata: "
+          << cpp_strerror(r) << dendl;
+      return r;
+    }
+    marker = mgr->get_marker(handle);
+
+    if (!keys.empty()) {
+      assert(keys.size() == 1);
+      auto& key = keys.front();
+      if (!callback(std::move(key), std::move(marker))) {
+        return 0;
+      }
+    }
+  } while (truncated);
+
+  if (start_marker.empty()) {
+    // already listed all keys
+    return 0;
+  }
+
+  // restart the listing from the beginning (empty marker)
+  mgr->list_keys_complete(handle);
+  handle = nullptr;
+
+  r = mgr->list_keys_init(section, "", &handle);
+  if (r < 0) {
+    ldout(cct, 10) << "failed to restart metadata listing: "
+        << cpp_strerror(r) << dendl;
+    return r;
+  }
+  ldout(cct, 20) << "restarting metadata listing" << dendl;
+
+  do {
+    // get the next key and marker
+    r = mgr->list_keys_next(handle, 1, keys, &truncated);
+    if (r < 0) {
+      ldout(cct, 10) << "failed to list metadata: "
+          << cpp_strerror(r) << dendl;
+      return r;
+    }
+    marker = mgr->get_marker(handle);
+
+    if (!keys.empty()) {
+      assert(keys.size() == 1);
+      auto& key = keys.front();
+      // stop at original marker
+      if (marker >= start_marker) {
+        return 0;
+      }
+      if (!callback(std::move(key), std::move(marker))) {
```

Comment:
what if caller does not exist anymore? E.g., CR manager was shut down. callback might not be valid I think.

Reply:
you're right, the lambda in `BucketTrimCR` is capturing a `this` pointer that may not be valid for the lifetime of `MetadataListCR`

it looks like this lambda can safely capture a reference to `this` via `boost::intrusive_ptr`. that way, the lambda capture will keep this `BucketTrimCR` alive until `MetadataListCR` destructs. because we spawn `MetadataListCR` here with a `call()`, it's the parent stack that holds a reference to the `MetadataListCR`, rather than `BucketTrimCR` itself - that means we don't have to worry about cyclical references

```diff
+        // capture a reference so 'this' remains valid in the callback
+        auto ref = boost::intrusive_ptr<RGWCoroutine>{this};
         // list cold buckets to consider for trim
-        auto cb = [this] (std::string&& bucket, std::string&& marker) {
+        auto cb = [this, ref] (std::string&& bucket, std::string&& marker) {
```

Reply:
addressed in https://github.com/ceph/ceph/pull/17761/commits/b86144215f38645afa750a0198c3a466fa7a6ee8 and verified with valgrind that it doesn't leak